LEAN-4341: LW-FY25-Q4-S1: Invalid Email with test@r form Accepted in …

nghazzawi-atypon opened the pull request:
- Added email format validation using regex.
- Prevent saving/updating when email is invalid
- Show inline error messages

mnatsheh commented:
Can't we use the `pattern` prop like the "ORCID" field?

nghazzawi-atypon commented:
> Can't we use the `pattern` prop like the "ORCID" field?

done @mnatsheh

mnatsheh commented on src/components/authors/AuthorDetailsForm.tsx:
```diff
@@ -139,19 +170,31 @@ export const AuthorDetailsForm: React.FC<AuthorDetailsFormProps> = ({
+                        hasError={!!error}
+                        {...props.field}
+                        ref={emailRef}
+                        onBlur={(e) => {
```
Do we need to explicitly validate on blur? shouldn't it happen automatically?

nghazzawi-atypon commented on src/components/authors/AuthorDetailsForm.tsx:
```diff
@@ -85,6 +102,20 @@ export const AuthorDetailsForm: React.FC<AuthorDetailsFormProps> = ({
   authorFormRef,
 }) => {
   const formRef = useRef<FormikProps<ContributorAttrs>>(null)
+  const emailRef = useRef<HTMLInputElement>(null)
```
removed

mnatsheh commented:
I'm actually confused by the entire implementation. Why is adding pattern="[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}" to the input not enough?

nghazzawi-atypon commented:
> I'm actually confused by the entire implementation. Why is adding pattern="[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+.[a-zA-Z]{2,}" to the input not enough?

the pattern attribute on an input provides basic client-side validation, it falls short for several reasons in this context. First, pattern is browser-dependent and lacks customizable error messaging, limiting user feedback to generic browser alerts (e.g., "Please match the requested format"). Second, it cannot enforce dynamic rules, such as requiring an email only for corresponding authors (isCorresponding), which our validateForm function handles via isEmailRequired. Third, pattern alone doesn’t integrate with React state management to disable the "Save" or "Update Details" buttons for invalid formats, a key requirement to prevent invalid data submission.


I have added more new changes to the PR included : 

1. **Introduced a new utility/lib file for form validation**:
   - Added a `fieldValidators.ts` file to centralize validation logic for name and email fields. This utility is designed as a scalable place to add validations for other form fields in the future.

2. **Introduced a new dialog for invalid format**:
   - Implemented a `ConfirmationDialog` component to handle invalid format cases (e.g., `DialogType.INVALID`).

3. **Disabled "Update Details" button for invalid fields and empty or required fields**:
   - Added logic to disable the "Update Details" button when fields are invalid (e.g., email format issues) or when required fields.

4. **Enhanced the use of `setTimeout`**:
   - Replaced standalone `setTimeout` in `handleSaveAuthor` with a `useEffect` hook tied to `lastSavedAuthor`, using `setTimeout` to clear the highlight after 3.2 seconds. Advantages include:
     - **Automatic Cleanup**: The `useEffect` cleanup function (`clearTimeout`) prevents memory leaks and React warnings.
     - **Controlled Timing**: Dependency on `lastSavedAuthor` ensures only one timer runs at a time, avoiding stacking from rapid saves.
     - Aligns with React’s declarative model, managing side effects safely within the component lifecycle.

5. **Prevented save or update details for invalid email format**:
   - Integrated `validateForm` to check email validity using `fieldValidators.ts`, blocking `handleSave` and "Update Details" actions when the email format is invalid, with appropriate dialog feedback.

6. **Performed code cleanup**:
   - Refactored `handleSaveAuthor` and `handleClose` to reduce duplication by centralizing validation and dialog logic in `showDialog` and `validateForm`.
   - Reused `normalize` function consistently to compare author states, simplifying change detection.
   - Removed unused `authorFormRef` reference and streamlined state management (e.g., `unSavedChanges`, `nextAuthor`) 
7. Consolidated Dialog State: Combined the three separate dialog states into a single dialogState object for better management.

mnatsheh commented on src/components/authors/AuthorsModal.tsx:
```diff
@@ -181,12 +186,18 @@ export const AuthorsModal: React.FC<AuthorsModalProps> = ({
-    showRequiredFieldConfirmationDialog,
-    setShowRequiredFieldConfirmationDialog,
-  ] = useState(false)
+  const [dialogState, setDialogState] = useState<{
```
Instead of the custom type here, we should just store `Partial<ConfirmationDialogProps>`. This way, we don't need to update it every time we add a new dialog. An `undefined` value means the dialog is not shown,

mnatsheh commented on src/components/authors/AuthorsModal.tsx:
```diff
@@ -221,15 +229,65 @@ export const AuthorsModal: React.FC<AuthorsModalProps> = ({
+  const [selection, setSelection] = useState(author)
+
+  // Centralized dialog handler
+  const showDialog = (
```
Same as above. We can just accept a general `Partial<ConfirmationDialogProps>` arg.

nghazzawi-atypon replied:
done

mnatsheh commented on src/components/authors/AuthorsModal.tsx:
```diff
@@ -221,15 +229,65 @@ export const AuthorsModal: React.FC<AuthorsModalProps> = ({
+  }
+
+  // Reusable form validation
+  const validateFormFields = (
+    values: ContributorAttrs,
+    requireEmail: boolean
+  ): { dialogType: 'required' | 'invalid' | 'none'; fieldName?: string } => {
```
This should only accept `values`. We should be able to figure out the value of `requireEmail` without it being passed.

nghazzawi-atypon replied:
done

mnatsheh commented on src/lib/fieldValidators.ts:
```diff
@@ -0,0 +1,71 @@
+}
+
+// Email validation function
+export const validateEmail = (
+  email: string,
+  isRequired = false
+): ValidationResult => {
+  if (isRequired && !email) {
+    return {
+      isValid: false,
+      error: FIELD_ERRORS.email.required,
+      field: 'email',
+    }
+  }
+  if (email && !emailRegex.test(email)) {
+    return { isValid: false, error: FIELD_ERRORS.email.invalid, field: 'email' }
+  }
+  return { isValid: true, field: 'email' }
+}
+
+// Name validation function for given-name and family-name
+export const validateName = (
+  given: string,
+  family: string
+): ValidationResult => {
+  if (!given || !family) {
+    return {
+      isValid: false,
+      error: FIELD_ERRORS.name.required,
+      field: 'name',
+    }
+  }
+  return { isValid: true, field: 'name' }
+}
```
I don't see much value in these to be honest. We should just incorporate the logic into `validateFormFields`.

nghazzawi-atypon replied:
done
